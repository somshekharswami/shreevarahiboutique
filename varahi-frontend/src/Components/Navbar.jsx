import React, { useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { CgShoppingBag } from "react-icons/cg";
import { GrCart } from "react-icons/gr";
import { FiUser, FiLogOut } from "react-icons/fi";
import { useAuth } from "../context/AuthContext";
import { signOut } from "firebase/auth";
import { auth } from "../firebase";

const Navbar = ({ searchItem, onSearchChange, onClearSearch }) => {
  const [isMobileMenuOpen, setIsMobileMenuOpen] = useState(false);
  const { currentUser, setCurrentUser } = useAuth();
  const navigate = useNavigate();

  const handleSearchSubmit = (e) => {
    if (e.key === "Enter" && searchItem.trim()) {
      navigate(`/search?q=${searchItem.trim()}`);
    }
  };

  const handleLogout = async () => {
    try {
      await signOut(auth);
      setCurrentUser(null);
      localStorage.removeItem("user");

      navigate("/login");
    } catch (error) {
      console.error("Logout failed", error);
    }
  };
  const closeMenu = () => setIsMobileMenuOpen(false);

  return (
    <nav className="sticky top-0 z-50 bg-pink-700 text-white shadow-md">
      <div className="max-w-7xl mx-auto px-4 sm:px-6 lg:px-8">
        <div className="flex justify-between items-center h-16">
          {/* Logo */}
          <Link
            to="/"
            className="text-[32px] md:text-[40px] font-serif font-semibold tracking-normal leading-none hover:text-pink-200 transition"
          >
            VB
          </Link>

          {/* Desktop Search */}
          <div className=" md:flex flex-1 justify-center px-4">
            <div className="relative w-full max-w-md">
              <input
                type="text"
                aria-label="searchBar"
                placeholder="kurti, dupatta, palazzo.."
                value={searchItem}
                onChange={(e) => onSearchChange(e.target.value)}
                className="w-full px-4 py-2 pr-10 rounded-lg outline-none text-black placeholder-gray-500 font-mono"
              />
              {searchItem && (
                <button
                  onClick={onClearSearch}
                  className="absolute right-3 top-1/2 transform -translate-y-1/2 text-gray-400 hover:text-red-500 text-xl font-bold"
                >
                  &times;
                </button>
              )}
            </div>
          </div>

          {/* Desktop Icons */}
          <div className="hidden md:flex space-x-6 items-center">
            {currentUser ? (
              <div className="flex items-center gap-2">
                <span>ðŸ‘— {currentUser.name}</span>
                <button
                  aria-label="logout"
                  onClick={handleLogout}
                  className="hover:text-red-400 transition"
                  title="Logout"
                >
                  <FiLogOut size={28} />
                </button>
              </div>
            ) : (
              <Link
                to="/login"
                aria-label="login"
                className="hover:text-pink-200 transition"
                title="Login"
              >
                <FiUser size={28} />
              </Link>
            )}

            <Link
              to="/my-orders"
              aria-label="My Orders"
              className="block text-white px-2 py-2 hover:text-pink-300"
            >
              <CgShoppingBag
                data-testid="bagIcon"
                size={30}
                className="inline mr-1"
              />
            </Link>

            <Link
              to="/cart"
              aria-label="cart-icon"
              className="block text-white px-1 py-1 hover:text-pink-300"
            >
              <GrCart
                data-testid="cartIcon"
                size={29}
                className="inline mr-1"
              />
            </Link>
          </div>

          {/* Mobile Hamburger */}
          <div className="md:hidden">
            <button
              aria-label="hamburgerbtn"
              onClick={() => setIsMobileMenuOpen(!isMobileMenuOpen)}
              className="focus:outline-none"
            >
              <svg
                className="w-6 h-6"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth={2}
                  d={
                    isMobileMenuOpen
                      ? "M6 18L18 6M6 6l12 12"
                      : "M4 6h16M4 12h16M4 18h16"
                  }
                />
              </svg>
            </button>
          </div>
        </div>

        {/* Mobile Menu Content */}
        {isMobileMenuOpen && (
          <div className="md:hidden space-y-4 pb-4">
            {currentUser && (
              <div className="flex items-center gap-2 text-lg px-2">
                ðŸ‘— {currentUser.name}
                <button
                  onClick={handleLogout}
                  className="hover:text-red-300 transition ml-auto"
                >
                  <FiLogOut size={24} />
                </button>
              </div>
            )}

            {!currentUser && (
              <Link
                to="/login"
                aria-label="login"
                onClick={closeMenu}
                className="block text-white px-2 hover:text-pink-300"
              >
                <FiUser className="inline mr-1" />
                Login
              </Link>
            )}
            <Link
              to="/my-orders"
              aria-label="my orders"
              onClick={closeMenu}
              className="block text-white px-2 hover:text-pink-300"
            >
              <CgShoppingBag className="inline mr-1" />
              My Orders
            </Link>

            <Link
              to="/cart"
              aria-label="cart-icon"
              onClick={closeMenu}
              className="block text-white px-2 hover:text-pink-300"
            >
              <GrCart className="inline mr-1" />
              Cart
            </Link>
            {/* Collections Dropdown - Mobile Only */}
            <div className="px-2">
              <details className="group">
                <summary className="flex justify-between items-center cursor-pointer text-white hover:text-pink-300 py-1 transition">
                  <span className="font-medium">Collections</span>
                  <svg
                    className="w-5 h-5 transition-transform duration-300 ease-in-out group-open:rotate-180"
                    fill="none"
                    stroke="currentColor"
                    viewBox="0 0 24 24"
                  >
                    <path
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth={2}
                      d="M19 9l-7 7-7-7"
                    />
                  </svg>
                </summary>
                <ul className="pl-2 mt-2 space-y-1 text-pink-100 max-h-64 overflow-y-auto scrollbar-hide">
                  <li>
                    <Link
                      to="/kurti"
                      aria-label="kurti"
                      onClick={closeMenu}
                      className="block px-2 py-1 rounded hover:bg-pink-600 active:scale-95 transition duration-200"
                    >
                      Kurti
                    </Link>
                  </li>
                  <li>
                    <Link
                      to="/duppata"
                      aria-label="duppata"
                      onClick={closeMenu}
                      className="block px-2 py-1 rounded hover:bg-pink-600 active:scale-95 transition duration-200"
                    >
                      Duppata
                    </Link>
                  </li>
                  <li>
                    <Link
                      to="/two-piece-suits"
                      aria-label="two-piece suits"
                      onClick={closeMenu}
                      className="block px-2 py-1 rounded hover:bg-pink-600 active:scale-95 transition duration-200"
                    >
                      Two-piece Suits
                    </Link>
                  </li>
                  <li>
                    <Link
                      to="/three-piece-suits"
                      onClick={closeMenu}
                      className="block px-2 py-1 rounded hover:bg-pink-600 active:scale-95 transition duration-200"
                    >
                      Three-piece Suits
                    </Link>
                  </li>
                  <li>
                    <Link
                      to="/leggings"
                      onClick={closeMenu}
                      className="block px-2 py-1 rounded hover:bg-pink-600 active:scale-95 transition duration-200"
                    >
                      Leggings
                    </Link>
                  </li>
                  <li>
                    <Link
                      to="/palazzo"
                      onClick={closeMenu}
                      className="block px-2 py-1 rounded hover:bg-pink-600 active:scale-95 transition duration-200"
                    >
                      Palazzo
                    </Link>
                  </li>

                  <li>
                    <Link
                      to="/pants"
                      onClick={closeMenu}
                      className="block px-2 py-1 rounded hover:bg-pink-600 active:scale-95 transition duration-200"
                    >
                      Pants
                    </Link>
                  </li>
                </ul>
              </details>
            </div>
          </div>
        )}
      </div>
    </nav>
  );
};

export default Navbar;
