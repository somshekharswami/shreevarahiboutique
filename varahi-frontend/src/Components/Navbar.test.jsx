import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import Navbar from "./Navbar";

// âœ… Mock AuthContext â€” use dynamic value for currentUser
const mockSetCurrentUser = jest.fn();

let mockAuthValue = {
  currentUser: null,
  setCurrentUser: mockSetCurrentUser,
  logout: jest.fn(),
};

jest.mock("../../src/context/AuthContext", () => ({
  useAuth: () => mockAuthValue,
}));

// âœ… Helper: Renders Navbar with React Router
const renderWithRouter = (initialRoute = "/") => {
  const mockOnSearchChange = jest.fn();
 return render(
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route
          path="/"
          element={<Navbar onSearchChange={mockOnSearchChange} />}
        />
        <Route path="/my-orders" element={<div>My Orders Page</div>} />
        <Route path="/cart" element={<div>Contact Page</div>} />
        <Route path="/login" element={<div>Login Page</div>} />
        <Route path="/kurti" element={<div>kurti page</div>} />
      </Routes>
    </MemoryRouter>
  );
};

describe("Navbar Component", () => {
  test("renders logo and is visible", () => {
    renderWithRouter();
    const logo = screen.getByRole("link", { name: /vb/i });
    expect(logo).toBeInTheDocument();
    expect(logo).toBeVisible();
  });

  test("renders searchBar, visibility, allows typing input", async () => {
    renderWithRouter();
    const searchBar = screen.getByLabelText(/searchBar/i);
    expect(searchBar).toBeInTheDocument();
    expect(searchBar).toBeVisible();
    await userEvent.type(searchBar, "kurti");
    expect(searchBar).toHaveValue("kurti");
  });

  test("shows login icon when user is logged out", () => {
    mockAuthValue.currentUser = null;
    renderWithRouter();
    const login = screen.getByLabelText(/login/i);
    expect(login).toBeInTheDocument();
    expect(login).toBeVisible();
  });

  test("renders shopping bag icon and navigates correctly", async () => {
    renderWithRouter();
    const shoppingBag = screen.getByLabelText(/my orders/i);
    const bagIcon = screen.getByTestId(/bagicon/i);

    expect(shoppingBag).toBeInTheDocument();
    expect(bagIcon).toBeVisible();

    await userEvent.click(shoppingBag);
    expect(screen.getByText(/my orders page/i)).toBeInTheDocument();
  });

  test("renders cart icon and is visible", async () => {
    renderWithRouter();
    const cartLink = screen.getByLabelText(/cart-icon/i);
    const cartIcon = screen.getByTestId(/carticon/i);

    expect(cartLink).toBeInTheDocument();
    expect(cartIcon).toBeVisible();

    await userEvent.click(cartLink);
    expect(screen.getByText(/contact page/i)).toBeInTheDocument();
  });

  test("shows user name and logout icon when user is logged in", () => {
    // âœ… Set up mock user BEFORE rendering
    mockAuthValue.currentUser = {
      name: "Somshekhar",
      uid: "123",
      email: "somshekhar@example.com",
    };

    renderWithRouter();
    //47 31 33
    expect(screen.getByText(/ðŸ‘— Somshekhar/i)).toBeInTheDocument();
    expect(screen.getByLabelText("logout")).toBeVisible();
    expect(screen.queryByLabelText("login")).not.toBeInTheDocument(); // âœ… This will now pass
  });

  // for Mobile menu Collection DropDown Tests
  test("Mobile menu 'Collections' All prodcuts link renders and navigates", async () => {
    renderWithRouter();

    // Opens the hamburger menu
    const hamburgerBtn = screen.getByLabelText(/hamburgerbtn/i);
    await userEvent.click(hamburgerBtn);

    const collectionsSummary = screen.getByText(/collections/i);
    await userEvent.click(collectionsSummary);

    //1. check for kurti
    const kurtiLink = screen.getByRole("link", { name: /kurti/i });
    expect(kurtiLink).toBeInTheDocument();
    expect(kurtiLink).toBeVisible();

    await userEvent.click(kurtiLink);
    expect(screen.getByText(/kurti page/i)).toBeInTheDocument();
  });
  test("matches snapshot", () => {
    const { asFragment } = renderWithRouter();
    expect(asFragment()).toMatchSnapshot();
  });
});
