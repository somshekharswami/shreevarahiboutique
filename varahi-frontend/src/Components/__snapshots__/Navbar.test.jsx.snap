// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Navbar Component matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="sticky top-0 z-50 bg-pink-700 text-white shadow-md"
  >
    <div
      class="max-w-7xl mx-auto px-4 sm:px-6 lg:px-8"
    >
      <div
        class="flex justify-between items-center h-16"
      >
        <a
          class="text-[32px] md:text-[40px] font-serif font-semibold tracking-normal leading-none hover:text-pink-200 transition"
          data-discover="true"
          href="/"
        >
          VB
        </a>
        <div
          class=" md:flex flex-1 justify-center px-4"
        >
          <div
            class="relative w-full max-w-md"
          >
            <input
              aria-label="searchBar"
              class="w-full px-4 py-2 pr-10 rounded-lg outline-none text-black placeholder-gray-500 font-mono"
              placeholder="kurti, dupatta, palazzo.."
              type="text"
            />
          </div>
        </div>
        <div
          class="hidden md:flex space-x-6 items-center"
        >
          <div
            class="flex items-center gap-2"
          >
            <span>
              ðŸ‘— Somshekhar
            </span>
            <button
              aria-label="logout"
              class="hover:text-red-400 transition"
              title="Logout"
            >
              <svg
                fill="none"
                height="28"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="28"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
                />
                <polyline
                  points="16 17 21 12 16 7"
                />
                <line
                  x1="21"
                  x2="9"
                  y1="12"
                  y2="12"
                />
              </svg>
            </button>
          </div>
          <a
            aria-label="My Orders"
            class="block text-white px-2 py-2 hover:text-pink-300"
            data-discover="true"
            href="/my-orders"
          >
            <svg
              class="inline mr-1"
              data-testid="bagIcon"
              fill="none"
              height="30"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="30"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M5 4H19C19.5523 4 20 4.44771 20 5V19C20 19.5523 19.5523 20 19 20H5C4.44772 20 4 19.5523 4 19V5C4 4.44772 4.44771 4 5 4ZM2 5C2 3.34315 3.34315 2 5 2H19C20.6569 2 22 3.34315 22 5V19C22 20.6569 20.6569 22 19 22H5C3.34315 22 2 20.6569 2 19V5ZM12 12C9.23858 12 7 9.31371 7 6H9C9 8.56606 10.6691 10 12 10C13.3309 10 15 8.56606 15 6H17C17 9.31371 14.7614 12 12 12Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </a>
          <a
            aria-label="cart-icon"
            class="block text-white px-1 py-1 hover:text-pink-300"
            data-discover="true"
            href="/cart"
          >
            <svg
              class="inline mr-1"
              data-testid="cartIcon"
              fill="currentColor"
              height="29"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="29"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5,5 L22,5 L20,14 L7,14 L4,2 L0,2 M7,14 L8,18 L21,18 M19,23 C18.4475,23 18,22.5525 18,22 C18,21.4475 18.4475,21 19,21 C19.5525,21 20,21.4475 20,22 C20,22.5525 19.5525,23 19,23 Z M9,23 C8.4475,23 8,22.5525 8,22 C8,21.4475 8.4475,21 9,21 C9.5525,21 10,21.4475 10,22 C10,22.5525 9.5525,23 9,23 Z"
                fill="none"
                stroke-width="2"
              />
            </svg>
          </a>
        </div>
        <div
          class="md:hidden"
        >
          <button
            aria-label="hamburgerbtn"
            class="focus:outline-none"
          >
            <svg
              class="w-6 h-6"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
            >
              <path
                d="M4 6h16M4 12h16M4 18h16"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
